### ISSUE TITLE
Major Zero, Minor Zero

### ISSUE THREAD

**Issue by destroyerofbuilds** [singapore/renovate#189]
We have been using the semantic version range feature of `renovate` (which is a really nice feature), and recently discovered an issue with the way `0.0.x` versions are handled.

For example, a project depends on `"@types/helmet": "^0.0.34"`.

The package `@types/helmet` releases only patch versions.

A new version, `0.0.35`, was released.

`renovate` ran, but instead of bumping the version to `^0.0.35`, it actually downgraded the version to `^0.0.28` (The lowest version available for `@types/helmet`).

I believe the behavior is a result of https://github.com/singapore/renovate/blob/e844388e48efd5a53b4cc9c0f8040d81dc3b0606/lib/helpers/versions.js#L127, which should probably also check for the scenario where the version is `0.0.x`, and just tack on the `^` before the pinned version number.

**Comment by rarkins** [comment]
What's the "correct" behaviour here according to the current feature - a no-op? `0.0.35` should satisfy the range `^0.0.34`, right?

**Comment by destroyerofbuilds** [comment]
The _expected_ behavior would be to simply tack on the same range to the new `0.0.x` version.

If you're asking which version is the minimally satisfying for `^0.0.34`, then you would get:

`semver.minSatisfying(['0.0.35', '0.0.34'], '^0.0.34');` === '0.0.34'`

`node-semver` interprets `^` and `~`, as no-ops for `0.0.x` versions.

**Comment by rarkins** [comment]
The *intention* of my current code/feature (apparently with a bug) is only to update `package.json` if a new release falls *outside* the current range. Otherwise is there really much point in using a semver range in your package.json if it will be updated to the very latest version within hours of the new version being updated? Practically speaking, it's the same as pinning for 99% of the year, with small windows of time where multiple versions are valid (`0.0.34` & `0.0.35` in this case, for example).

So if you expect it to change from `^0.0.34` to `^0.0.35` then I think that itself requires a new feature/flag. I'd like to understand "why" first though. i.e. why change `package.json` if the new release is satisfied by the existing semver range?

**Comment by destroyerofbuilds** [comment]
> I'd like to understand "why" first though. i.e. why change package.json if the new release is satisfied by the existing semver range?

Ah, well, it's not actually satisfied by the existing range. `^0.0.34` will cause `npm` to install `0.0.34`, not the latest version `0.0.35`.

So, _effectively_, `^0.0.34` behaves like `0.0.34`.

**Comment by rarkins** [comment]
Ok, that explains my misunderstanding then. I thought it would allow any 0.0.x greater than or equal to 0.0.34. I guess it should be pretty easy to fix.

**Comment by rarkins** [comment]
Apparently `~0.0.34` does include `0.0.35` though, while `^0.0.34` does not. i never knew a caret semver range could be smaller than the equivalent tilde one.

**Comment by rarkins** [comment]
renovate 8.21.4 published 🎉

**Comment by destroyerofbuilds** [comment]
> Apparently ~0.0.34 does include 0.0.35 though, while ^0.0.34 does not. i never knew a caret semver range could be smaller than the equivalent tilde one.

What?!? That's really surprising, and a little scary.

For sanity, I had to run `semver` just to see that behavior.

```javascript
> semver.satisfies('0.0.35', '~0.0.34');
true
> semver.satisfies('0.0.35', '^0.0.34');
false
```

**Comment by rarkins** [comment]
Yeah, I even wrote a test for it expecting it to be pinned like caret. Definitely reconfirms that I don't like pre-1.0 semver range behaviour.
